logger/logger_types.go

```diff
@@ -2,34 +2,90 @@ package logger
 
 import (
 	"io"
-	"log"
-	"os"
+	"sync"
 )
 
-// Logging struct that holds all user configurable options for the logger
-type Logging struct {
-	Enabled              *bool  `json:"enabled,omitempty"`
-	File                 string `json:"file"`
-	ColourOutput         bool   `json:"colour"`
-	ColourOutputOverride bool   `json:"colourOverride,omitempty"`
-	Level                string `json:"level"`
-	Rotate               bool   `json:"rotate"`
+const timestampFormat = " 02/01/2006 15:04:05 "
+
+const spacer = "|"
+
+type Config struct {
+	Enabled          *bool            `json:"enabled"`
+	AdvancedSettings advancedSettings `json:"advancedSettings"`
+	SubLoggers       []SubLoggers     `json:"subloggers"`
 }
 
-var (
-	debugLogger *log.Logger
-	infoLogger  *log.Logger
-	warnLogger  *log.Logger
-	errorLogger *log.Logger
-	fatalLogger *log.Logger
+type headers struct {
+	Info  string `json:"info"`
+	Warn  string `json:"warn"`
+	Debug string `json:"debug"`
+	Error string `json:"error"`
+}
+
+type advancedSettings struct {
+	Spacer          string  `json:"spacer"`
+	TimeStampFormat string  `json:"timeStampFormat"`
+	Headers         headers `json:"headers"`
+}
 
-	logFileHandle *os.File
+type SubLoggers struct {
+	Name   string `json:"name"`
+	Level  string `json:"level"`
+	Output string `json:"output"`
+}
 
-	logOutput io.Writer
+type Logger struct {
+	Timestamp                                        string
+	InfoHeader, ErrorHeader, DebugHeader, WarnHeader string
+	Spacer                                           string
+}
+
+type subLogger struct {
+	Info, Debug, Warn, Error bool
+	output                   io.Writer
+}
 
-	// LogPath location to store logs in
+type LogEvent struct {
+	data   []byte
+	output io.Writer
+}
+
+type multiWriter struct {
+	writers []io.Writer
+	mu      sync.Mutex
+}
+
+var (
+	logger           = &Logger{}
+	GlobalLogConfig  = &Config{}
+	subSystemLoggers = map[string]subLogger{}
+	eventPool        = &sync.Pool{
+		New: func() interface{} {
+			return &LogEvent{
+				data: make([]byte, 0, 80),
+			}
+		},
+	}
 	LogPath string
+)
+
+const (
+	LogGlobal = "log"
+
+	SubSystemConnMgr = "connection"
```

Comment:
This is the minor of the minoriest nits. I feel that these don't need the "SubSystem" prefix. Any confusion could be prevented by comments. Since a global/subsystem is required for each log, then the user will understand that the specific log type is a subsytem on account of not being called Global. This isn't enough for me to request changes though and don't mind if you prefer the prefix and it stays because I also like things being explicit 😆